### DISCUSSION TITLE
`type mismatch in closure argument`: Compile error when adding a wrapped submodule

### DISCUSSION THREAD

**Discussion by twitu** [PyO3/pyo3#3399] [Questions]
I have a module given below that compiles properly

```rust
#[pymodule]
pub fn common(_: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_class::<LogGuard>()?;
    m.add_function(wrap_pyfunction!(set_global_log_collector, m)?)?;
    Ok(())
}
```

However when I try to add this module as a submodule to another module it gives a type related error.

```rust
    let sys = PyModule::import(py, "sys")?;
    let sys_modules: &PyDict = sys.getattr("modules")?.downcast()?;

    // Common
    let submodule = pyo3::wrap_pymodule!(nautilus_common::common);
    m.add_wrapped(submodule)?;
    sys_modules.set_item(
        "nautilus_trader.core.nautilus_pyo3.common",
        m.getattr("common")?,
    )?;
```

```
 error[E0631]: type mismatch in closure arguments
Error:    --> pyo3/src/lib.rs:61:19
    |
60  |     let submodule = pyo3::wrap_pymodule!(nautilus_common::common);
    |                     --------------------------------------------- found signature defined here
61  |     m.add_wrapped(submodule)?;
    |       ----------- ^^^^^^^^^ expected due to this
    |       |
    |       required by a bound introduced by this call
    |
    = note: expected closure signature `fn(pyo3::Python<'_>) -> _`
               found closure signature `fn(pyo3::marker::Python<'_>) -> _`
note: required by a bound in `pyo3::prelude::PyModule::add_wrapped`
   --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.19.2/src/types/module.rs:300:56
    |
300 |     pub fn add_wrapped<'a, T>(&'a self, wrapper: &impl Fn(Python<'a>) -> T) -> PyResult<()>
    |                                                        ^^^^^^^^^^^^^^^^^^^ required by this bound in `PyModule::add_wrapped`
```

It appears to be a very minor difference in types `pyo3::Python<'_>` and `pyo3::marker::Python<'_>`. I'm not sure what's wrong here because I'm using similar code to add other submodules as well which does work without any issue.

Has anybody encountered this before?

**Comment by davidhewitt** [comment]
This sort of error usually occurs because there are two crates called `pyo3` in use. Maybe a name conflict?

**Reply by twitu** [reply]
But I have only one dependency for pyo3 in the workspace `Cargo.lock`. Cargo tree too only shows one pyo3.

```
[[package]]
name = "pyo3"
version = "0.19.2"
```

**Reply by davidhewitt** [reply]
`pyo3/src/lib.rs:61:19`

- is your crate itself called pyo3?

**Comment by twitu** [comment]
Oh actually that's just the directory name. The library is named is set to
`nautilus_pyo3` in Cargo.toml. Surely the directory name doesn't cause
conflict because I'd wrapped other submodules earlier and those worked.

The only difference between this wrapped module and the others is that this
one adds a function. Others only add classes.

**Reply by davidhewitt** [reply]
Right I think it's probably fine. To offer more insight I think I'd need you to send a minimal reproduction.
